File: rust/equity-options-engine/src/monte_carlo.rs
//! Monte Carlo pricing of European options under GBM (exact scheme).
//!
//! The terminal stock price is simulated exactly:
//! `S_T = S exp((r - q - sigma^2/2) T + sigma sqrt(T) Z)`, `Z ~ N(0, 1)`,
//! so there is *no* time-discretisation bias — only statistical error,
//! which is reported as a standard error and 95% confidence interval.
//!
//! # Variance reduction (semantics mirror the Python reference)
//!
//! * **Antithetic variates**: pairs `(Z, -Z)`; the standard error is
//!   computed on the *pair averages* (the correct estimator for
//!   correlated pairs).
//! * **Control variate**: the discounted terminal stock `exp(-rT) S_T`
//!   with known mean `S exp(-qT)` (martingale property); the optimal
//!   coefficient is estimated from the sample covariance.
//!
//! Every entry point takes an explicit `u64` seed feeding the in-crate
//! SplitMix64 -> xoshiro256++ generator ([`crate::rng`]). Same seed =>
//! **bit-identical** result on every platform.
//!
//! Conventions: continuously compounded annualised `r`, `q` (ACT/365F),
//! `t` in years, `sigma` annualised.

use crate::black_scholes::{bs_price, validate_inputs, OptionType, PricingError};
use crate::rng::Xoshiro256PlusPlus;

/// Two-sided 95% normal quantile (matches the Python reference constant).
const Z95: f64 = 1.959_963_984_540_054;

/// Monte Carlo estimate with statistical error bars.
///
/// # Examples
///
/// ```
/// use eq_options_engine::{mc_price, OptionType};
/// let res = mc_price(100.0, 100.0, 1.0, 0.05, 0.2, 0.0, OptionType::Call,
///                    50_000, true, true, 42).unwrap();
/// assert!(res.std_error > 0.0);
/// assert!(res.ci_low < res.price && res.price < res.ci_high);
/// assert!(res.contains(10.450583572185565)); // analytic BS well inside CI
/// ```
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct McResult {
    /// Point estimate of the present value, currency units.
    pub price: f64,
    /// Standard error of the estimator.
    pub std_error: f64,
    /// Lower edge of the two-sided 95% confidence interval.
    pub ci_low: f64,
    /// Upper edge of the two-sided 95% confidence interval.
    pub ci_high: f64,
    /// Number of simulated paths (antithetic pairs count as 2 paths).
    pub n_paths: usize,
}

impl McResult {
    /// Return `true` if `x` lies inside the 95% confidence interval.
    ///
    /// # Examples
    ///
    /// ```
    /// use eq_options_engine::McResult;
    /// let r = McResult { price: 10.0, std_error: 0.1,
    ///                    ci_low: 9.804, ci_high: 10.196, n_paths: 1000 };
    /// assert!(r.contains(10.1));
    /// assert!(!r.contains(11.0));
    /// ```
    pub fn contains(&self, x: f64) -> bool {
        self.ci_low <= x && x <= self.ci_high
    }
}

/// Build an [`McResult`] from i.i.d. per-draw samples (ddof = 1).
fn summary(samples: &[f64], n_paths: usize) -> McResult {
    let n = samples.len();
    let mean = samples.iter().sum::<f64>() / (n as f64);
    let se = if n > 1 {
        let var = samples
            .iter()
            .map(|x| {
                let d = x - mean;
                d * d
            })
            .sum::<f64>()
            / ((n - 1) as f64);
        (var / (n as f64)).sqrt()
    } else {
        0.0
    };
    McResult {
        price: mean,
        std_error: se,
        ci_low: mean - Z95 * se,
        ci_high: mean + Z95 * se,
        n_paths,
    }
}

/// Monte Carlo price of a European option under exact-scheme GBM.
///
/// # Arguments
///
/// * `s`, `k`, `t`, `r`, `sigma`, `q`, `option_type` — as in
///   [`bs_price`].
/// * `n_paths` — total number of paths (rounded up to even when
///   `antithetic`), `>= 2`.
/// * `antithetic` — use antithetic pairs `(Z, -Z)`.
/// * `control_variate` — use the discounted terminal stock as a control
///   variate with the sample-optimal coefficient.
/// * `seed` — explicit RNG seed; same seed => bit-identical result.
///
/// `t == 0` or `sigma == 0` are deterministic; the exact Black-Scholes
/// value is returned with `std_error = 0`.
///
/// # Errors
///
/// [`PricingError::InvalidInput`] on invalid inputs or `n_paths < 2`.
///
/// # Examples
///
/// ```
/// use eq_options_engine::{mc_price, OptionType};
/// let a = mc_price(100.0, 100.0, 1.0, 0.05, 0.2, 0.0, OptionType::Call,
///                  10_000, true, true, 7).unwrap();
/// let b = mc_price(100.0, 100.0, 1.0, 0.05, 0.2, 0.0, OptionType::Call,
///                  10_000, true, true, 7).unwrap();
/// assert_eq!(a.price.to_bits(), b.price.to_bits()); // bit-reproducible
/// ```
#[allow(clippy::too_many_arguments)]
pub fn mc_price(
    s: f64,
    k: f64,
    t: f64,
    r: f64,
    sigma: f64,
    q: f64,
    option_type: OptionType,
    n_paths: usize,
    antithetic: bool,
    control_variate: bool,
    seed: u64,
) -> Result<McResult, PricingError> {
    validate_inputs(s, k, t, sigma)?;
    if n_paths < 2 {
        return Err(PricingError::InvalidInput(format!(
            "n_paths must be >= 2, got {n_paths}"
        )));
    }
    if t == 0.0 || sigma == 0.0 {
        let exact = bs_price(s, k, t, r, sigma, q, option_type)?;
        return Ok(McResult {
            price: exact,
            std_error: 0.0,
            ci_low: exact,
            ci_high: exact,
            n_paths,
        });
    }

    let mut rng = Xoshiro256PlusPlus::new(seed);
    let disc = (-r * t).exp();
    let sign = option_type.sign();
    let drift = (r - q - 0.5 * sigma * sigma) * t;
    let vol_sqrt_t = sigma * t.sqrt();
    let terminal = |z: f64| s * (drift + vol_sqrt_t * z).exp();

    // Draw normals: fresh draws first, then (for antithetic) their mirrors,
    // matching the Python layout `concatenate([z, -z])`.
    let (z, n_eff): (Vec<f64>, usize) = if antithetic {
        let half = n_paths.div_ceil(2);
        let mut z = vec![0.0_f64; 2 * half];
        let (fresh, mirror) = z.split_at_mut(half);
        rng.fill_standard_normal(fresh);
        for i in 0..half {
            mirror[i] = -fresh[i];
        }
        (z, 2 * half)
    } else {
        let mut z = vec![0.0_f64; n_paths];
        rng.fill_standard_normal(&mut z);
        (z, n_paths)
    };

    let mut payoff: Vec<f64> = z
        .iter()
        .map(|&zi| disc * (sign * (terminal(zi) - k)).max(0.0))
        .collect();

    if control_variate {
        // Control: discounted terminal stock, known mean S exp(-qT).
        let control: Vec<f64> = z.iter().map(|&zi| disc * terminal(zi)).collect();
        let control_mean = s * (-q * t).exp();
        let n = n_eff as f64;
        let mean_p = payoff.iter().sum::<f64>() / n;
        let mean_c = control.iter().sum::<f64>() / n;
        let mut cov_pc = 0.0;
        let mut var_c = 0.0;
        for i in 0..n_eff {
            let dp = payoff[i] - mean_p;
            let dc = control[i] - mean_c;
            cov_pc += dp * dc;
            var_c += dc * dc;
        }
        cov_pc /= n - 1.0;
        var_c /= n - 1.0;
        let beta = if var_c > 0.0 { cov_pc / var_c } else { 0.0 };
        for i in 0..n_eff {
            payoff[i] -= beta * (control[i] - control_mean);
        }
    }

    if antithetic {
        let half = n_eff / 2;
        let samples: Vec<f64> = (0..half)
            .map(|i| 0.5 * (payoff[i] + payoff[half + i]))
            .collect();
        Ok(summary(&samples, n_eff))
    } else {
        Ok(summary(&payoff, n_eff))
    }
}

File: cpp/equity-options-engine/src/monte_carlo.cpp
#include "eqopt/monte_carlo.hpp"

#include <cmath>
#include <cstddef>
#include <random>
#include <stdexcept>
#include <thread>
#include <vector>

#include "eqopt/black_scholes.hpp"

namespace eqopt {

namespace {

constexpr double kZ95 = 1.959963984540054;  // two-sided 95% normal quantile

/// Acklam's rational approximation to the inverse normal CDF (~1.15e-9
/// relative error), used as the initial guess before Halley refinement.
double inv_norm_cdf_acklam(double p) noexcept {
    // Coefficients from P. J. Acklam (2003).
    constexpr double a[6] = {-3.969683028665376e+01, 2.209460984245205e+02,
                             -2.759285104469687e+02, 1.383577518672690e+02,
                             -3.066479806614716e+01, 2.506628277459239e+00};
    constexpr double b[5] = {-5.447609879822406e+01, 1.615858368580409e+02,
                             -1.556989798598866e+02, 6.680131188771972e+01,
                             -1.328068155288572e+01};
    constexpr double c[6] = {-7.784894002430293e-03, -3.223964580411365e-01,
                             -2.400758277161838e+00, -2.549732539343734e+00,
                             4.374664141464968e+00,  2.938163982698783e+00};
    constexpr double d[4] = {7.784695709041462e-03, 3.224671290700398e-01,
                             2.445134137142996e+00, 3.754408661907416e+00};
    constexpr double p_low = 0.02425;

    if (p < p_low) {
        const double t = std::sqrt(-2.0 * std::log(p));
        return (((((c[0] * t + c[1]) * t + c[2]) * t + c[3]) * t + c[4]) * t +
                c[5]) /
               ((((d[0] * t + d[1]) * t + d[2]) * t + d[3]) * t + 1.0);
    }
    if (p <= 1.0 - p_low) {
        const double u = p - 0.5;
        const double t = u * u;
        return (((((a[0] * t + a[1]) * t + a[2]) * t + a[3]) * t + a[4]) * t +
                a[5]) *
               u /
               (((((b[0] * t + b[1]) * t + b[2]) * t + b[3]) * t + b[4]) * t +
                1.0);
    }
    const double t = std::sqrt(-2.0 * std::log(1.0 - p));
    return -(((((c[0] * t + c[1]) * t + c[2]) * t + c[3]) * t + c[4]) * t +
             c[5]) /
           ((((d[0] * t + d[1]) * t + d[2]) * t + d[3]) * t + 1.0);
}

/// splitmix64 mix step — derives statistically independent per-thread seeds
/// from (base seed, chunk index) deterministically.
std::uint64_t splitmix64(std::uint64_t x) noexcept {
    x += 0x9e3779b97f4a7c15ULL;
    x = (x ^ (x >> 30)) * 0xbf58476d1ce4e5b9ULL;
    x = (x ^ (x >> 27)) * 0x94d049bb133111ebULL;
    return x ^ (x >> 31);
}

}  // namespace

double normal_from_u64(std::uint64_t u64) noexcept {
    // 53-bit uniform in (0, 1): (k + 0.5) * 2^-53 never hits 0 or 1.
    const double u =
        (static_cast<double>(u64 >> 11) + 0.5) * 1.1102230246251565e-16;
    double x = inv_norm_cdf_acklam(u);
    // One Halley step against the erfc-based CDF: pushes the ~1e-9 error of
    // the rational approximation down to ~1e-15.
    const double e = norm_cdf(x) - u;
    const double f = norm_pdf(x);
    x -= e / (f + 0.5 * x * e);
    return x;
}

MCResult mc_price(double S, double K, double T, double r, double sigma,
                  double q, OptionType type, std::size_t n_paths,
                  bool antithetic, bool control_variate, std::uint64_t seed,
                  unsigned threads) {
    validate_inputs(S, K, T, sigma);
    validate_rates(r, q);
    if (n_paths < 2) {
        throw std::invalid_argument("n_paths must be >= 2");
    }
    if (threads == 0) {
        throw std::invalid_argument("threads must be >= 1");
    }
    if (T == 0.0 || sigma == 0.0) {
        const double exact = bs_price(S, K, T, r, sigma, q, type);
        return MCResult{exact, 0.0, exact, exact, n_paths};
    }

    const double disc = std::exp(-r * T);
    const double sign = type == OptionType::Call ? 1.0 : -1.0;
    const double drift = (r - q - 0.5 * sigma * sigma) * T;
    const double vol_sqrt_t = sigma * std::sqrt(T);

    // `base` independent draws; with antithetic each is mirrored to -Z.
    const std::size_t half = antithetic ? (n_paths + 1) / 2 : 0;
    const std::size_t base = antithetic ? half : n_paths;
    const std::size_t n_eff = antithetic ? 2 * half : n_paths;

    std::vector<double> payoff(n_eff);
    std::vector<double> control;
    if (control_variate) control.resize(n_eff);

    // Deterministic chunking: chunk c covers [c*base/W, (c+1)*base/W) with
    // its own RNG stream seeded from splitmix64(seed, c). For a fixed
    // (seed, threads) the draws — and therefore every output bit — are
    // identical run-to-run.
    const unsigned n_workers = static_cast<unsigned>(
        std::min<std::size_t>(threads, base));
    auto worker = [&](unsigned c) {
        const std::size_t lo = base * c / n_workers;
        const std::size_t hi = base * (c + 1) / n_workers;
        std::mt19937_64 rng(splitmix64(seed + 0x51ed2701ULL * c));
        for (std::size_t i = lo; i < hi; ++i) {
            const double z = normal_from_u64(rng());
            const double s_up = S * std::exp(drift + vol_sqrt_t * z);
            payoff[i] = disc * std::fmax(sign * (s_up - K), 0.0);
            if (control_variate) control[i] = disc * s_up;
            if (antithetic) {
                const double s_dn = S * std::exp(drift - vol_sqrt_t * z);
                payoff[half + i] = disc * std::fmax(sign * (s_dn - K), 0.0);
                if (control_variate) control[half + i] = disc * s_dn;
            }
        }
    };
    if (n_workers == 1) {
        worker(0);
    } else {
        std::vector<std::thread> pool;
        pool.reserve(n_workers);
        for (unsigned c = 0; c < n_workers; ++c) pool.emplace_back(worker, c);
        for (auto& t : pool) t.join();
    }

    if (control_variate) {
        // Sample-optimal coefficient beta = cov(payoff, control)/var(control).
        double mean_p = 0.0;
        double mean_c = 0.0;
        for (std::size_t i = 0; i < n_eff; ++i) {
            mean_p += payoff[i];
            mean_c += control[i];
        }
        mean_p /= static_cast<double>(n_eff);
        mean_c /= static_cast<double>(n_eff);
        double cov_pc = 0.0;
        double var_c = 0.0;
        for (std::size_t i = 0; i < n_eff; ++i) {
            const double dp = payoff[i] - mean_p;
            const double dc = control[i] - mean_c;
            cov_pc += dp * dc;
            var_c += dc * dc;
        }
        cov_pc /= static_cast<double>(n_eff - 1);
        var_c /= static_cast<double>(n_eff - 1);
        const double beta = var_c > 0.0 ? cov_pc / var_c : 0.0;
        const double control_mean = S * std::exp(-q * T);  // martingale mean
        for (std::size_t i = 0; i < n_eff; ++i) {
            payoff[i] -= beta * (control[i] - control_mean);
        }
    }

    // Antithetic pairs are correlated: build i.i.d. samples from pair means.
    std::size_t n_samples;
    const double* samples;
    std::vector<double> pair_means;
    if (antithetic) {
        pair_means.resize(half);
        for (std::size_t i = 0; i < half; ++i) {
            pair_means[i] = 0.5 * (payoff[i] + payoff[half + i]);
        }
        samples = pair_means.data();
        n_samples = half;
    } else {
        samples = payoff.data();
        n_samples = n_eff;
    }

    double mean = 0.0;
    for (std::size_t i = 0; i < n_samples; ++i) mean += samples[i];
    mean /= static_cast<double>(n_samples);
    double var = 0.0;
    for (std::size_t i = 0; i < n_samples; ++i) {
        const double d = samples[i] - mean;
        var += d * d;
    }
    const double se =
        n_samples > 1
            ? std::sqrt(var / static_cast<double>(n_samples - 1) /
                        static_cast<double>(n_samples))
            : 0.0;
    return MCResult{mean, se, mean - kZ95 * se, mean + kZ95 * se, n_eff};
}

}  // namespace eqopt
